extensions/amp-analytics/0.1/events.js

```diff
@@ -306,6 +334,90 @@ export class EventTracker {
   add(unusedContext, unusedEventType, unusedConfig, unusedListener) {}
 }
 
+/**
+ * Tracks custom browser events.
+ */
+export class CustomBrowserEventTracker extends EventTracker {
+  /**
+   * @param {!./analytics-root.AnalyticsRoot} root
+   */
+  constructor(root) {
+    super(root);
+
+    /** @private {?Observable<!Event>} */
+    this.observables_ = new Observable();
+
+    /** @private {?function(!Event)} */
+    this.boundOnSession_ = this.observables_.fire.bind(this.observables_);
+
+    /** @private {?function(!Event)} */
+    this.debouncedBoundOnSession_ = debounce(
+      this.root.ampdoc.win,
+      this.boundOnSession_,
+      TIME_WAIT
+    );
+  }
+
+  /** @override */
+  dispose() {
+    const root = this.root.getRoot();
+    Object.keys(BrowserEventType).forEach((key) => {
+      root.removeEventListener(
+        BrowserEventType[key],
+        this.debouncedBoundOnSession_
+      );
+    });
+    this.boundOnSession_ = null;
+    this.observables_ = null;
+  }
+
+  /** @override */
+  add(context, eventType, config, listener) {
+    userAssert(
+      isExperimentOn(this.root.ampdoc.win, 'analytics-browser-events'),
+      'expected global "analytics-browser-events" experiment to be enabled'
+    );
+
+    const {
+      'on': eventName,
+      'selectionMethod': selectionMethod = null,
+      'selector': selector,
+    } = config;
+
+    userAssert(
+      selector.length,
+      'Missing required selector on browser event trigger'
+    );
+    assertUniqueSelectors(selector);
+
+    const targetPromises = this.root.getElements(
+      context,
+      selector,
+      selectionMethod,
+      false
+    );
+
+    this.root
+      .getRootElement()
+      .addEventListener(eventName, this.debouncedBoundOnSession_, true);
+
```

Comment:
What happens if we have two `'on': 'blur' rules? That means that we would trigger `debouncedBoundOnSession_` twice, right? Could we keep track of which events we're already listening to?

Reply:
Yes it triggers it twice. To avoid duplicate listener for the same event, I have made an object of the browser events with value initally set to false. Once a listener is added to the event, the value will be set to true. Because there is already a listener, further attempts to add a listener won't be successful. 
There's also an alternate solution of overriding `Element.addEventListener` to remember the added listeners, but I believe that the former is a better solution.

Reply:
Thanks for explaining. That solution sounds good to me, just made a few comments about its details.